PowerEditor/src/MISC/Common/Common.cpp

```diff
@@ -1333,23 +1333,33 @@ int nbDigitsFromNbLines(size_t nbLines)
 
 generic_string getDateTimeStrFrom(const generic_string& dateTimeFormat, const SYSTEMTIME& st)
 {
-	generic_string dateTimeStr = dateTimeFormat;
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("Y"), std::to_wstring(st.wYear));
-	wchar_t buf[3];
-	_snwprintf(buf, sizeof(buf), TEXT("%02d"), st.wMonth);
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("M"), buf);
+	const TCHAR* localeName = LOCALE_NAME_USER_DEFAULT;
+	const DWORD flags = 0;
 
-	_snwprintf(buf, sizeof(buf), TEXT("%02d"), st.wDay);
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("D"), buf);
+	const int bufferSize = MAX_PATH;
+	TCHAR buffer[bufferSize] = {};
 
-	_snwprintf(buf, sizeof(buf), TEXT("%02d"), st.wHour);
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("h"), buf);
+	GetTimeFormatEx(localeName, flags, &st, dateTimeFormat.c_str(), buffer, bufferSize);
+	GetDateFormatEx(localeName, flags, &st, buffer, buffer, bufferSize, nullptr);
 
-	_snwprintf(buf, sizeof(buf), TEXT("%02d"), st.wMinute);
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("m"), buf);
+	const int M_numberSize = 4;
+	TCHAR M_number[M_numberSize] = {};
+	GetDateFormatEx(localeName, flags, &st, TEXT("M"), M_number, M_numberSize, nullptr);
 
-	_snwprintf(buf, sizeof(buf), TEXT("%02d"), st.wSecond);
-	dateTimeStr = stringReplace(dateTimeStr, TEXT("s"), buf);
+	// "tt" will be translated to "AM" or "PM" by GetTimeFormatEx(),
+	// then 'M' in "AM" or "PM" will be transformed to a number 1-12 by GetDateFormatEx().
+	// So if user provides "tt", it will be A1-A12 or P1-P12 in the buffer.
+	// Here we search A1-A12 or P1-P12 and restore them back to AM or PM.
+	generic_string dateTimeStr = buffer;
+
+	generic_string AM_number = TEXT("A");
+	AM_number += M_number;
+
+	generic_string PM_number = TEXT("P");
+	PM_number += M_number;
+
+	dateTimeStr = stringReplace(dateTimeStr, AM_number, TEXT("AM"));
+	dateTimeStr = stringReplace(dateTimeStr, PM_number, TEXT("PM"));
 

```

Comment:
Maybe instead of postprocessing, a bit of preprocessing will be a better option:
```c++
	generic_string formatEx = dateTimeFormat;
	size_t tPos = 0;
	while ((tPos = formatEx.find(L't', tPos)) != std::string::npos)
	{
		size_t tEnd = tPos + 1;
		for ( ; tEnd < formatEx.length() && formatEx[tEnd] == L't'; ++tEnd);
		formatEx.insert(tEnd, 4, L'\'');
		formatEx.insert(tPos, 4, L'\'');
		tPos += tEnd + 8;
	}
	GetTimeFormatEx(localeName, flags, &st, formatEx.c_str(), buffer, bufferSize);

	formatEx = buffer;
	GetDateFormatEx(localeName, flags, &st, formatEx.c_str(), buffer, bufferSize, nullptr);
```
This way no matter into what `t` or `tt` is converted, it will always be escaped for `GetDateFormatEx()`.

Reply:
Amazing @ivan-u7n !
The test input:
`hh:mm:ss, tt; M dd yy g t`
transformation:
```
hh:mm:ss, ''''tt''''; M dd yy g ''''t''''
01:24:00, 'AM'; M dd yy g 'A'
```
Expected output: 
`01:24:00, AM; 10 26 85 A.D. A`

My question is: how did you know the trick of `''''xxx''''` for `GetTimeFormatEx`? Any ref about that?
Why `GetDateFormatEx` remove only 1 pair of quotes instead of 3?

Reply:
BTW, may I integrate your code or you prefer to create a PR under your name?

Reply:
The "trick" is from https://docs.microsoft.com/en-us/windows/win32/intl/hour--minute--and-second-format-pictures:
> Single quotation marks can be used to mark characters that should be displayed exactly as specified. If the application must display a single quotation mark, it should place two single quotation marks in a row. For example, 'abc''bar', is displayed as "abc'bar".

So to get a single quotation mark, there are must be 4 of them: two at the ends are for the start and the end of the escaped string, and two in the middle are turned into one in the output.

However the code above will break if a time marker is absent or there's a single quotation mark in it. That code also doesn't do anything to user escaped strings. This one will only break if there a single quotation mark in a time marker:
```c++
generic_string getDateTimeStrFrom(const generic_string& dateTimeFormat, const SYSTEMTIME& st)
{
	const TCHAR* localeName = LOCALE_NAME_USER_DEFAULT;
	const DWORD flags = 0;

	const int bufferSize = MAX_PATH;
	TCHAR buffer[bufferSize] = {};

	generic_string formatEx = dateTimeFormat;
	size_t findPos = 0;
	size_t tResultLength = std::string::npos;
	while ((findPos = formatEx.find_first_of(L"'t", findPos)) != std::string::npos)
	{
		if (formatEx[findPos] == L'\'')
		{
			// single quotation mark escape
			size_t findEnd = findPos + 1;
			while ((findEnd = formatEx.find_first_of(L"'", findEnd)) != std::string::npos)
			{
				if (findEnd + 1 < formatEx.length() && formatEx[findEnd + 1] == L'\'')
				{
					formatEx.insert(findEnd, 2, L'\'');
					findEnd += 4;
					continue;
				}
				break;
			}
			if (findEnd == std::string::npos)
			{
				// no closing single quotation mark
				formatEx.push_back(L'\'');
				findEnd = formatEx.length();
			}
			formatEx.insert(findEnd, 2, L'\'');
			formatEx.insert(findPos, 2, L'\'');
			findPos = findEnd + 4 + 1;
		}
		else
		{
			// time marker
			if (tResultLength == std::string::npos)
			{
				// if time marker is defined in locale...
				GetTimeFormatEx(localeName, flags, &st, L"t", buffer, bufferSize);
				tResultLength = wcslen(buffer);
			}
			size_t findEnd = formatEx.find_first_not_of(L"t", findPos + 1);
			if (findEnd == std::string::npos)
				findEnd = formatEx.length();
			if (tResultLength)
			{
				// ...then escape its result...
				formatEx.insert(findEnd, 4, L'\'');
				formatEx.insert(findPos, 4, L'\'');
				findPos = findEnd + 8;
			}
			else
			{
				// ...or remove it: its somewhat magical removal by GetTimeFormatEx() breaks escaping
				formatEx.erase(findPos, findEnd - findPos);
			}
		}
	}
	GetTimeFormatEx(localeName, flags, &st, formatEx.c_str(), buffer, bufferSize);

	formatEx = buffer;
	GetDateFormatEx(localeName, flags, &st, formatEx.c_str(), buffer, bufferSize, nullptr);

	return buffer;
}
```
`hh:mm:'s''s' tt; 'yy''yy'-MM-dd` => `07:48:s's ; yy'yy-09-15`

Should I try to fix the last small bug (the breakage if there a single quotation mark in a time marker) and then open a PR for my version?

Reply:
Is this code easier than my [PR](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10480)?

Reply:
> However the code above will break if a time marker is absent or there's a single quotation mark in it. That code also doesn't do anything to user escaped strings

Example please ?

>. This one will only break if there a single quotation mark in a time marker:

Example please ?

Reply:
@mere-human 
> Is this code easier than my [PR](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10480)?

This one yes:
https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10549#discussion_r709139701

Reply:
@ivan-u7n 
> Should I try to fix the last small bug (the breakage if there a single quotation mark in a time marker) and then open a PR for my version?

The code becomes more complex than the first version you provided.
If the most cases work with the first version, it'll be fine.

Reply:
Small summary of the latest approaches:

This one [#10549 (comment)](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10549#discussion_r709139701) will not work if the AM/PM symbols are absent/empty in the windows settings.

This one [#10480 (comment)](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10480#issuecomment-920032850) will cover the cases of empty AM/PM symbols and is not much more complicated. So it is the more reliable solution.

If single quotation mark should be handled in customised AM/PM symbols, the [#10480 approach](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10480) of @mere-human looks like the way to go.

If more escaping should be handled, everything gets complex ;-)

Reply:
> If more escaping should be handled, everything gets complex ;-)

No overly complex, only ~100 lines of code: #10556 😎